lib/message_request.js

```diff
@@ -0,0 +1,167 @@
+var logger = require('minilog')('message:request');
+
+var opTable = {
+  control: ['nameSync'],
+  message: ['publish', 'subscribe', 'sync', 'unsubscribe'],
+  presence: ['get', 'set', 'subscribe', 'sync', 'unsubscribe'],
+  status: ['get', 'set', 'subscribe', 'sync', 'unsubscribe'],
+  stream: ['get', 'push', 'subscribe', 'sync', 'unsubscribe']
+};
+
+var Request = function (message) {
+  this.message = message;
+
+  if (!this._isValid()) {
+    logger.error('invalid request. op: ' + this.op + '; to: ' + this.to);
+  }
+};
+
+Request.buildGet = function (scope, options) {
+  var message = { op: 'get', to: scope};
+  return new Request(message).setOptions(options);
+};
+
+Request.buildPublish = function (scope, value) {
+  var message = { op: 'publish', to: scope};
+  var request = new Request(message);
+  request.setAttr('value', value);
+
+  return request;
+};
+
+Request.buildPush = function (scope, resource, action, value) {
+  var message = { op: 'push', to: scope};
+  var request = new Request(message);
+  request.setAttr('resource', resource);
+  request.setAttr('action', action);
+  request.setAttr('value', value);
+
+  return request;
+};
+
+Request.buildNameSync = function (scope, options) {
+  var message = { op: 'nameSync', to: scope};
+  return new Request(message).setOptions(options);
+};
+
+Request.buildSet = function (scope, value, key, userType, clientData) {
+  var message = { op: 'set', to: scope};
+  var request = new Request(message);
+  request.setAttr('value', value);
+  request.setAttr('key', key);
+  request.setAttr('type', userType);
+  if (clientData) {
+    request.setAttr('clientData', clientData);
+  }
+
+  return request;
+};
+
+Request.buildSync = function (scope, options) {
+  var message = { op: 'sync', to: scope};
+  return new Request(message).setOptions(options);
+};
+
+Request.buildSubscribe = function (scope, options) {
+  var message = { op: 'subscribe', to: scope};
+  return new Request(message).setOptions(options);
+};
+
+Request.buildUnsubscribe = function (scope, options) {
+  var message = { op: 'unsubscribe', to: scope};
+  return new Request(message);
+};
+
+// Instance methods
+
+Request.prototype.setAuthData = function (configuration) {
+  this.setAttr('userData', configuration.userData);
+  if (configuration.auth) {
+    this.setAttr('auth', configuration.auth);
+    this.setAttr('userId', configuration.userId);
+    this.setAttr('userType', configuration.userType);
+    this.setAttr('accountName', configuration.accountName);
+  }
+};
+
+Request.prototype.getMessage = function () {
+  return this.message;
+};
+
+Request.prototype.setOptions = function (options) {
+  if (options) {
+    this.setAttr('options', options);
+  } else if (!options && this.isPresence() && this.getAttr('op') === 'sync') {
+    this.setAttr('options', { version: 2});
+  }
```

Comment:
This special case is in the wrong place. I would prefer this to be done this way:

```
Request.prototype.buildSync = function(options) {
  var message = { op: 'sync', to: scope};
  var request = new Request(message).setOptions(options);
  if(request.isPresence()) {
    request.forceV2Sync();
  }
  return request;
}

Request.prototype.forceV2Sync = function() {
  this.options ||= {};
  this.options.version = 2;
}
```

This way, the code is easy to read, and we will find it easy to remove.